controllers/binding/snapshotenvironmentbinding_adapter.go

```diff
@@ -141,3 +141,72 @@ func (a *Adapter) getDeploymentTargetForEnvironment(environment *applicationapiv
 
 	return deploymentTarget, nil
 }
+
+// EnsureEphemeralEnvironmentsCleanedUp will ensure that ephemeral environment(s) are cleaned up.
+func (a *Adapter) EnsureEphemeralEnvironmentsCleanedUp() (reconciler.OperationResult, error) {
+	testEnvironment := a.environment
+
+	isEphemeral := false
+	for _, tag := range testEnvironment.Spec.Tags {
+		if tag == "ephemeral" {
+			isEphemeral = true
+			break
+		}
+	}
+
+	if isEphemeral {
+		dtc, err := a.loader.GetDeploymentTargetClaimForEnvironment(a.client, a.context, testEnvironment)
+		if err != nil || dtc == nil {
+			a.logger.Error(err, "Failed to find deploymentTargetClaim defined in environment", "environment.Name", testEnvironment.Name)
+			return reconciler.RequeueWithError(err)
+		}
+
+		dt, err := a.loader.GetDeploymentTargetForDeploymentTargetClaim(a.client, a.context, dtc)
+		if err != nil || dt == nil {
+			a.logger.Error(err, "Failed to find deploymentTarget defined in deploymentTargetClaim", "deploymentTargetClaim.Name", dtc.Name)
+			return reconciler.RequeueWithError(err)
+		}
+
+		binding, err := a.loader.FindExistingSnapshotEnvironmentBinding(a.client, a.context, a.application, testEnvironment)
+		if err != nil || binding == nil {
+			a.logger.Error(err, "Failed to find snapshotEnvironmentBinding associated with environment", "environment.Name", testEnvironment.Name)
+			return reconciler.RequeueWithError(err)
+		}
+
+		a.logger.Info("Deleting deploymentTarget", "deploymentTarget.Name", dt.Name)
+		err = a.client.Delete(a.context, dt)
+		if err != nil {
+			a.logger.Error(err, "Failed to delete the deploymentTarget")
+			return reconciler.RequeueWithError(err)
+		}
+		a.logger.LogAuditEvent("DeploymentTarget deleted", dt, h.LogActionDelete)
+
+		a.logger.Info("Deleting deploymentTargetClaim", "deploymentTargetClaim.Name", dtc.Name)
+		err = a.client.Delete(a.context, dtc)
+		if err != nil {
+			a.logger.Error(err, "Failed to delete the deploymentTargetClaim")
+			return reconciler.RequeueWithError(err)
+		}
+		a.logger.LogAuditEvent("DeploymentTargetClaim deleted", dtc, h.LogActionDelete)
+
+		a.logger.Info("Deleting environment", "environment.Name", testEnvironment.Name)
+		err = a.client.Delete(a.context, testEnvironment)
+		if err != nil {
+			a.logger.Error(err, "Failed to delete the test ephemeral environment")
+			return reconciler.RequeueWithError(err)
+		}
+		a.logger.LogAuditEvent("Ephemeral environment deleted", testEnvironment, h.LogActionDelete)
+
+		a.logger.Info("Deleting snapshotEnvironmentBinding", "binding.Name", binding.Name)
```

Comment:
Note that deleting snapshotEnvironmentBinding should not be done after #236 is merged as the owner reference from the environment will ensure that the SEB is also deleted. It could be a good idea to create a function that handles cleanup of the environment as we talked about earlier so we don't duplicate the logic and miss things like this in the future.